File: apps/sepuling-java/src/main/java/smecalculus/bezmen/storage/SepulkaStateEm.java
package smecalculus.bezmen.storage;

import java.time.LocalDateTime;
import java.util.UUID;
import lombok.Data;
import org.springframework.data.annotation.Id;
import org.springframework.data.domain.Persistable;
import org.springframework.data.relational.core.mapping.Column;
import org.springframework.data.relational.core.mapping.Table;

public abstract class SepulkaStateEm {
    @Data
    public static class Existence {
        UUID internalId;
        String externalId;
    }

    @Data
    public static class Preview {
        UUID internalId;
        String externalId;
        LocalDateTime createdAt;
    }

    @Data
    public static class Touch {
        Integer revision;
        LocalDateTime updatedAt;
    }

    @Data
    @Table("sepulkas")
    public static class AggregateRoot implements Persistable<UUID> {
        @Id
        UUID internalId;

        @Column
        String externalId;

        @Column
        Integer revision;

        @Column
        LocalDateTime createdAt;

        @Column
        LocalDateTime updatedAt;

        @Override
        public UUID getId() {
            return internalId;
        }

        @Override
        public boolean isNew() {
            return true;
        }
    }
}

File: apps/sepuling-kotlin/src/main/kotlin/smecalculus/bezmen/storage/SepulkaStateEm.kt
package smecalculus.bezmen.storage

import org.springframework.data.annotation.Id
import org.springframework.data.domain.Persistable
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import java.time.LocalDateTime
import java.util.UUID

abstract class SepulkaStateEm {
    data class Existence(
        var internalId: UUID,
        var externalId: String,
    )

    data class Preview(
        var internalId: UUID,
        var externalId: String,
        var createdAt: LocalDateTime,
    )

    data class Touch(
        var revision: Int,
        var updatedAt: LocalDateTime,
    )

    @Table("sepulkas")
    data class AggregateRoot(
        @Id
        var internalId: UUID?,
        @Column
        var externalId: String?,
        @Column
        var revision: Int?,
        @Column
        var createdAt: LocalDateTime?,
        @Column
        var updatedAt: LocalDateTime?,
    ) : Persistable<UUID> {
        override fun getId(): UUID? {
            return internalId
        }

        override fun isNew(): Boolean {
            return true
        }
    }
}
